### ISSUE TITLE
Incorrect handling of formula levels?

### ISSUE THREAD

**Issue by kape1395** [tlaplus/tlapm#188]
I encountered a proof step that passed too easily.
Here is my attempt to reproduce the unexpectedly passed proof.

TL;DR: See `Spec_Inv2`.

It looks like it happens only if another module is involved; hence, FiniteSetTheorems is used here.

```tla
---- MODULE ERR ----
EXTENDS TLAPS, FiniteSetTheorems

VARIABLE v
vars == <<v>>
Inv1 == v = {}
Inv2 == IsFiniteSet(v)
Init == v = {}
Next == v' = {}
Spec == Init /\ [][Next]_vars

THEOREM FirstAttempt ==
    ASSUME
        NEW p,
        NEW n
    PROVE
        p /\ n => p'
PROOF OBVIOUS \* Because p is assumed to be CONSTANT?

THEOREM Spec_Inv1 ==
    Spec => Inv1
PROOF
    <1>1. Init => Inv1 BY DEF Init, Inv1
    <1>2. Inv1 /\ [Next]_vars => Inv1'
        <2>1. SUFFICES ASSUME Inv1, Next PROVE Inv1' BY DEF vars, Inv1
        \* <2>q. QED BY <2>1 \* FAIL - as expected
        \* <2>q. QED BY <2>1 DEF Next \* FAIL - as expected
        <2>q. QED BY <2>1 DEF Next, Inv1 \* PASS - as expected.
    <1>q. QED BY <1>1, <1>2, PTL DEF Spec

THEOREM Spec_Inv2 ==
    Spec => Inv2
PROOF
    <1>1. Init => Inv2 BY FS_EmptySet DEF Init, Inv2
    <1>2. Inv2 /\ [Next]_vars => Inv2'
        <2>1. SUFFICES ASSUME Inv2, Next PROVE Inv2' BY DEF vars, Inv2
        <2>q. QED BY <2>1 \* ERROR here. Should not be proven!
    <1>q. QED BY <1>1, <1>2, PTL DEF Spec

====
```

1. `FirstAttempt` is proved by OBVIOUS. It surprised me, but I guess it passes because `p` is considered a CONSTANT level symbol.
2. `Spec_Inv1` uses the invariant defined in this module. Thus, all work correctly.
3. `Spec_Inv2` involves a definition from FiniteSets, and the step `<2>q` should fail, but it passes.

The obligation of the unexpectedly proven step `<2>q` is shown as

```tla
ASSUME NEW VARIABLE v,
       /\ Inv2
       /\ Next 
PROVE  Inv2'
```

I also checked it on the TLAPM built from `main` as `tlapm --nofp ERR.tla`.

@muenchnerkindl, could you double-check it? I hope it is just my confusion.

**Comment by kape1395** [comment]
A bit smaller example demonstrating strange behavior:
```tla
---- MODULE ERR2 ----
EXTENDS FiniteSets

O1(x) == IsFiniteSet(x)
O2(x) == TRUE
O3(x) == x = x
O4(x) == Cardinality(x) = 0

VARIABLE v
I1 == O1(v)
I2 == O2(v)
I3 == O3(v)
I4 == O4(v)

LEMMA ASSUME I1 PROVE I1' OBVIOUS \* ERROR: Proved, but should not. No symbols are expanded.
LEMMA ASSUME I2 PROVE I2' OBVIOUS \* ERROR: Proved, despite O(_) is not expanded.
LEMMA ASSUME I3 PROVE I3' OBVIOUS \* OK: Failed, as expected.
LEMMA ASSUME I4 PROVE I4' OBVIOUS \* OK: Failed, as expected.

====
```

  - For `O2`, I can guess TLAPM deduces that the variable is not used, and thus, the operator is at the constant level.
  - It's strange that an operator using `Cardinality` works as expected, but `IsFiniteSet` does not.

**Comment by kape1395** [comment]
@muenchnerkindl, could you check if the fix in https://github.com/tlaplus/tlapm/pull/189 makes the results reasonable?
The fix itself is trivial and should not break anything. I would like to have it merged sooner than later.
This bug affects bounded `\A`, `\E`, and probably other constructs that introduce bound variables with a domain where the domain is non-constant.
